web/src/composebox_typeahead.ts

```diff
@@ -1336,12 +1341,102 @@ export function initialize({
         $element: $("input#stream_message_recipient_topic"),
         type: "input",
     };
+
     new Typeahead(stream_message_typeahead_input, {
         source(): string[] {
-            return topics_seen_for(compose_state.stream_id());
+            const topics = topics_seen_for(compose_state.stream_id());
+            const topic_input = stream_message_typeahead_input.$element.val()?.toString() ?? "";
+
+            let dm_option: string[] = [];
+            if (topic_input.length >= 3) {
+                const matched_users: User[] = people
+                    .get_realm_users()
+                    .filter(build_person_matcher(topic_input));
+
+                const exact_matches = matched_users.filter(
+                    (user) => user.full_name.toLowerCase() === topic_input.toLowerCase(),
+                );
+
+                if (exact_matches.length === 1) {
+                    const matched_user = exact_matches[0];
+                    user_schema.parse(matched_user);
+                    dm_option = [
+                        JSON.stringify({
+                            type: "dm_suggestion",
+                            user: matched_user,
+                        }),
+                    ];
+                }
+            }
+
+            return [...dm_option, ...topics];
         },
-        items: max_num_items,
+
+        updater(item: string): string | undefined {
+            let parsed_item: unknown;
+
+            try {
+                parsed_item = JSON.parse(item);
+            } catch {
+                return item;
+            }
```

Comment:
Why is this doing JSON parsing?

Reply:
I guess maybe the better question is why you're using JSON rather than objects? I don't see an obvious reason you need to push elements of other types into strings, just because topics are strings.